parser/charset/encoding.go

```diff
@@ -140,8 +141,40 @@ func (e *Encoding) EncodeInternal(dest, src []byte) []byte {
 	return dest
 }
 
+// validUTF8 checks whether there are illegal utf8 characters in []byte.
+func (e *Encoding) validUTF8(src []byte) ([]byte, error) {
+	resultBytes := src
+	for len(src) > 0 {
+		r, size := utf8.DecodeRune(src)
+		if r == utf8.RuneError && size == 1 {
+			return resultBytes, e.generateErr(src, characterLengthUTF8(src))
+		}
+		src = src[size:]
+	}
+
+	return resultBytes, nil
+}
+
+// validUTF8String checks whether there are illegal utf8 characters in string.
+func (e *Encoding) validUTF8String(src string) (string, error) {
+	resultStr := src
+	for len(src) > 0 {
+		r, size := utf8.DecodeRuneInString(src)
+		if r == utf8.RuneError && size == 1 {
+			srcBytes := []byte(src)
+			return resultStr, e.generateErr(srcBytes, characterLengthUTF8(srcBytes))
+		}
+		src = src[size:]
+	}
+
+	return resultStr, nil
+}
+
 // Decode convert bytes from a specific charset to utf-8 charset.
 func (e *Encoding) Decode(dest, src []byte) ([]byte, error) {
+	if e.name == encodings[CharsetUTF8].name {
```

Comment:
@tangenta PTAL, Do we need to check the utf8 here?

Reply:
The reason is https://github.com/pingcap/tidb/issues/29685